File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int answer = 0;

        for(int i=0;i< grid.length;i++) {

            for(int j=0;j<grid[0].length;j++) {
                if(grid[i][j] == '1') {
                    dfsMarking(grid,i,j);
                    answer++;
                }
            }
        }
        return answer;
    }
    
    public void dfsMarking(char[][] g, int i, int j) {
        /* 범위 외 혹은 땅이 아니면 종료 */
        if(i < 0 || j < 0 || i >= g.length || j >= g[0].length || g[i][j] != '1') return;
        /* 방문한 땅은 물로 마킹 */
        g[i][j] = '0';
        /* 다른 방향으로 재탐색 */
        dfsMarking(g,i+1,j);
        dfsMarking(g,i-1,j);
        dfsMarking(g,i,j+1);
        dfsMarking(g,i,j-1);
    }
    
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        res = 0
        m, n = len(grid), len(grid[0])

        visited = set()
        dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        queue = []
        def bfs():
            while queue:
                x, y = queue.pop()
                visited.add((x, y))
                print(x,y)
                for dir_x, dir_y in dirs:
                    new_x, new_y = x + dir_x, y + dir_y
                    if (0 <= new_x < m and 0 <= new_y < n and 
                        grid[new_x][new_y] == "1" and 
                        (new_x, new_y) not in visited):
                        queue.append((new_x, new_y))
                        visited.add((new_x, new_y))
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and (i, j) not in visited:
                    queue.append((i,j))
                    bfs()
                    res += 1

        return res
        